src/plugins/intel_cpu/src/nodes/scaled_attn.cpp

```diff
@@ -0,0 +1,806 @@
+                       present_value.stride(2),
+                       0.f,
+                       has_out_transpose ? &output_emb.at({b, m_start, h * head_size}) : &output_emb.at({b, h, m_start}),
+                       has_out_transpose ? output_emb.stride(1) : output_emb.stride(2),
+                       1);
+        });
+    }
+};
+#endif
+
+// 2nd token case : only 1 token in query
+template <typename RT>
+struct MHA_1Token {
+    PlainTensor<float> m_attn_w;
+    PlainTensor<float> m_temp;
+
+    MHA_1Token() : m_attn_w(true), m_temp(true) {}
+
+    PlainTensor<uint8_t> causal_mask;
+    bool select_nfltmax_at_0;  // set attn_score to -FLT_MAX when causal_mask[...] equal to this
+    void set_causal_mask(PlainTensor<uint8_t> mask, bool _select_nfltmax_at_0) {
+        causal_mask = mask;
+        select_nfltmax_at_0 = _select_nfltmax_at_0;
+    }
+
+    // Q, K, V is ready, do attention
+    // query         [B, H, q_len, S]
+    // present_key   [B, H, kv_len, S]  stride of last dim maybe > 1
+    // present_value [B, H, kv_len, S]
+    // attention_mask [B, 1, q_len, kv_len]
+    // alibi
+    // output_emb    [B, L1, H*S]
+    void operator()(PlainTensor<RT>& query,
+                    PlainTensor<RT>& present_key,
+                    PlainTensor<RT>& present_value,
+                    const PlainTensor<float>& alibi_mask,
+                    const PlainTensor<float>& attention_mask,
+                    PlainTensor<RT>& output_emb,
+                    const PlainTensor<int32_t>& beams,
+                    bool has_out_transpose,
+                    bool auto_causal,
+                    float d_scale = 0.0f) {
+        auto B = query.size(0);
+        auto H = query.size(1);
+        auto q_len = query.size(2);
+        auto S = query.size(3);
+        auto kv_len = present_key.size(2);
+        auto h_group_num = present_key.size(1);
+        size_t h_each_group_len = 0;
+        if (h_group_num != H) {
+            h_each_group_len = H / h_group_num;
+        }
+
+        if (d_scale == 0.0f)
+            d_scale = 1.0f / sqrt(S);
+
+        // use per-token kernel, for each k,v token
+        //  attn mask is a matrix of q_len(kv_len)
+        m_attn_w.resize({B, H, q_len, kv_len});
+
+        if (h_each_group_len) {
+            parallel_for3d(B, H / h_each_group_len, kv_len, [&](size_t b, size_t h_group, size_t pk) {
+                // which batch item should be used at postion pk?
+                auto b_kv = beams ? beams.at({b, pk}) : b;
+                for (size_t pq = 0; pq < q_len; pq++) {
+                    for (size_t h = h_group * h_each_group_len; h < (h_group + 1) * h_each_group_len; h++) {
+                        auto sum = attn_dot_product(&query.at({b, h, pq, 0}),
+                                                          &present_key.at({b_kv, h_group, pk, 0}),
+                                                        S,
+                                            precision_of<RT>::value);
```

Comment:
the kernel `attn_dot_product` is small and light-weighted, current implementation using `cross_compiled_file` tools actually will make a dispatcher wrapper which calls different version according to CPU id, and this is done at each call, which maybe an overhead for such small kernel.

Reply:
Move the loop to `attn_dot_product`.